Source/Core/Core/HW/GBACore.cpp

```diff
@@ -0,0 +1,517 @@
+  m_event.callback = [](mTiming* timing, void* context, u32 cycles_late) {
+    Core* core = static_cast<Core*>(context);
+    static_cast<::GBA*>(core->m_core->board)->earlyExit = true;
+    core->m_waiting_for_event = false;
+  };
+  m_event.priority = 0x80;
+}
+
+void Core::SendJoybusCommand(u64 gc_ticks, u8* buffer)
+{
+  if (!m_running)
+    return;
+
+  Command command{};
+  command.ticks = gc_ticks;
+  command.sync_only = buffer == nullptr;
+  if (buffer)
+    std::copy(buffer, buffer + 5, command.buffer.begin());
+
+  if (m_thread)
+  {
+    std::lock_guard<std::mutex> lock(m_queue_mutex);
+    m_command_queue.push(command);
+    m_idle = false;
+    m_command_cv.notify_one();
+  }
+  else
+  {
+    RunCommand(command);
+  }
+}
+
+std::vector<u8> Core::GetJoybusResponse()
+{
+  if (!m_running)
+    return {};
+
+  if (m_thread)
+  {
+    std::unique_lock<std::mutex> lock(m_response_mutex);
+    m_response_cv.wait(lock, [&] { return m_response_ready; });
+  }
+  m_response_ready = false;
+  return m_response;
+}
+
+void Core::Flush()
+{
+  if (!m_running || !m_thread)
+    return;
+  std::unique_lock<std::mutex> lock(m_queue_mutex);
+  m_response_cv.wait(lock, [&] { return m_idle; });
+}
+
+void Core::ThreadLoop()
+{
+  Common::SetCurrentThreadName(fmt::format("GBA{}", m_device_number + 1).c_str());
+  std::unique_lock<std::mutex> queue_lock(m_queue_mutex);
+  while (true)
+  {
+    m_command_cv.wait(queue_lock, [&] { return !m_command_queue.empty() || m_exit_loop; });
+    if (m_exit_loop)
+      break;
+    Command command{m_command_queue.front()};
+    m_command_queue.pop();
+    queue_lock.unlock();
+
+    RunCommand(command);
+
+    queue_lock.lock();
+    if (m_command_queue.empty())
+      m_idle = true;
+    m_response_cv.notify_one();
+  }
+}
+
+void Core::RunCommand(Command& command)
+{
+  RunUntil(command.ticks);
+  if (!command.sync_only)
+  {
+    u64 run_ahead = 0;
+    m_response.clear();
+    if (m_link_enabled)
+    {
+      int recvd = GBASIOJOYSendCommand(
+          &m_sio_driver, static_cast<GBASIOJOYCommand>(command.buffer[0]), &command.buffer[1]);
+      std::copy(command.buffer.begin() + 1, command.buffer.begin() + 1 + recvd,
+                std::back_inserter(m_response));
+      run_ahead = m_response.size() * SystemTimers::GetTicksPerSecond() / BYTES_PER_SECOND;
+    }
+
+    if (m_thread && !m_response_ready)
+    {
+      std::lock_guard<std::mutex> response_lock(m_response_mutex);
+      m_response_ready = true;
+      m_response_cv.notify_one();
+    }
+    else
+    {
+      m_response_ready = true;
+    }
+
+    if (run_ahead)
+      RunFor(run_ahead);
+  }
+}
+
+void Core::RunUntil(u64 gc_ticks)
+{
+  const u64 gc_frequency = SystemTimers::GetTicksPerSecond();
+  mTimingSchedule(
+      m_core->timing, &m_event,
+      static_cast<s32>((gc_ticks - m_last_gc_ticks) * GBA_ARM7TDMI_FREQUENCY / gc_frequency));
+  m_waiting_for_event = true;
+
+  s32 start_time = mTimingCurrentTime(m_core->timing);
```

Comment:
You may want to use `mTimingGlobalTime` which is a u64.

Reply:
Turns out that is only available when USE_DEBUGGERS is defined, this is not an issue since we frequently sync devices and already use an s32 for the event anyway